src/Npgsql/NpgsqlWriteBuffer.cs

```diff
@@ -257,7 +280,7 @@ internal async Task DirectWrite(ReadOnlyMemory<byte> memory, bool async)
             try
             {
                 if (async)
-                    await Underlying.WriteAsync(memory);
+                    await Underlying.WriteAsync(memory, cancellationToken);
```

Comment:
This adds cancellation to direct writes - that's great. Since we interact with the underlying stream, though, we need to reimplement the timeout mechanism as well. It's completely fine to do this outside of this PR (which is already big enough) - I'm adding a note in #3166.

Note #2444 which may no longer be relevant, as I see we were already doing async direct writes. I'll take a look at this later, don't worry about it.